feat: update interfaces guide. (#63)
zapatacomputing/orquestra-core: docs/guides/interfaces.rst

.. _interfaces:
========================
Interfaces
========================

> One of the biggest strengths of Orquestra is its modularity. Integrating new backends, optimizers, ansatzes, etc. does not require changing the core code - it requires only creating a new module that conforms to existing interfaces and therefore can be used across the whole platform. Using our interfaces help you add whatever building blocks you need for your project and ensure that they will work well with other existing elements.

Reviewer: Typo help -> helps

```suggestion
One of the biggest strengths of Orquestra is its modularity. Integrating new backends, optimizers, ansatzes, etc. does not require changing the core code - it requires only creating a new module that conforms to existing interfaces and therefore can be used across the whole platform. Using our interfaces helps you add whatever building blocks you need for your project and ensure that they will work well with other existing elements.
```

What is an interface?
=====================

> Before we get into the details, let's explain what we mean by "interfaces" in this context. We use this word as a shorthand for "a way to implement certain concepts in such a way, that as long as two implementations conform to a particular interface, they can be swapped and the program should still work correctly". Obviously, various implementations might work differently - e.g.: one simulator might be faster than the other - but they should be undistinguishable when it comes to inputs, outputs and how they are being used.

Reviewer: I think that the correct word is `indistinguishable`, see e.g. discussion [here](https://english.stackexchange.com/questions/3544/undistinguishable-vs-indistinguishable): 

```suggestion
Before we get into the details, let's explain what we mean by "interfaces" in this context. We use this word as a shorthand for "a way to implement certain concepts in such a way, that as long as two implementations conform to a particular interface, they can be swapped and the program should still work correctly". Obviously, various implementations might work differently - e.g.: one simulator might be faster than the other - but they should be indistinguishable when it comes to inputs, outputs and how they are being used.
```

Benefits of interfaces
======================

Interface offers the ability to define behaviour that can be implemented by unrelated classes without forcing them to share a common class hierarchy. This results in code flexibility and faster development with easier testing options. Other benefits include:

* Interoperability
* Modularity
* Dependencies decoupling

Interface implementation
========================

Currently interfaces in Orquestra are defined in the following 3 main libraries:

* `orquestra-quantum <https://github.com/zapatacomputing/orquestra-quantum>`_
* `orquestra-opt <https://github.com/zapatacomputing/orquestra-opt>`_
* `orquestra-vqa <https://github.com/zapatacomputing/orquestra-vqa>`_

> The interfaces are located in the ``api`` directory within each packages. They are implemented either as an ``abstract class`` or ``protocols``. `Abstract classes <https://docs.python.org/3/library/abc.html>`_ are classes that contain one or more abstract method. However, they do not contain any implementation. Subclasses are required to provide implementations for abstract methods. Abstract classes can also be considered blueprint classes. 

Reviewer: I'm going to hell for pointing this out, but... Abstract class CAN contain an implementation of an abstract method, they only force the method to be overridden in the child class. Anyway, nothing prevents us from using base implementation.  The following example demonstrates this behaviour (please don't do it in any production code):

```python
from abc import ABC, abstractmethod

class A(ABC):

    @abstractmethod
    def do_stuff(self):
        return 37

class B(A):

    def do_stuff(self):
        return super().do_stuff() + 2100

if __name__ == '__main__':
    obj = B()
    print(obj.do_stuff())
```

Having said that, I would reword this sentence to better reflect what ABC are about, maybe something like:
```suggestion
The interfaces are located in the ``api`` directory within each packages. They are implemented either as an ``abstract class`` or ``protocols``. `Abstract classes <https://docs.python.org/3/library/abc.html>`_ are classes that contain one or more abstract methods and thus cannot be instantiated. Concrete implementations (subclasses) of a given abstract base class have to override all of the abstract methods. Abstract base classes often serve as blueprint classes, defining common behaviours in terms of some specialized method(s) that has to be implemented in child classes.
```
?

However, using abstract classes requires using inheritance, which sometimes is undesirable. Therefore, we also make use of `Protocols <https://peps.python.org/pep-0544/>`_ . The main differences are that we don't have to explicitly express that particular implementation conforms to a protocol and that protocols work not just with classes, but also functions. 

> Interfaces provided by Orquestra are:

Reviewer: I think that the following sounds better, but it is very subjective so feel free to ignore it:
```suggestion
Orquestra provides the following interfaces:
```

* :class:`Optimizer <orquestra.opt.api.Optimizer>` (:ref:`guide <optimizers_guide>`)
* :class:`CostFunction <orquestra.opt.api.CostFunction>` (:ref:`guide <cost_function_guide>`)
* :class:`Ansatz <orquestra.vqa.api.ansatz.Ansatz>` (:ref:`guide <ansatzes_guide>`)
* :class:`Estimators <orquestra.quantum.api.estimation.EstimateExpectationValues>` (:ref:`guide <estimators_guide>`)

Passing parameters
------------------

You might wonder - what if I have two optimizers which take different sets of parameters? For example, you might want to use `L-BFGS-B <https://docs.scipy.org/doc/scipy/reference/optimize.minimize-lbfgsb.html#optimize-minimize-lbfgsb>`_ and `COBYLA <https://docs.scipy.org/doc/scipy/reference/optimize.minimize-cobyla.html#optimize-minimize-cobyla>`_ from `scipy`. This is how you would use them in scipy: 

.. literalinclude:: ../examples/interfaces_guide.py
    :start-after: >> Guide code snippet: script showing scipy optimizers
    :lines: 2-3
    :language: python

`L-BFGS-B` doesn't support argument called `constraints` and `COBYLA` doesn't support `bounds`. So how can you make them fit the same interface, if they have different inputs?

> The way we solved this is that in case of Optimizers, we created an interface which has method `minimize`, which requires providing only `cost_function` and `initial_params`. All the extra parameters are passed during the initialization of the Optimizer class, like this:

Reviewer: WDYT about rewording it as follows:

```suggestion
In the case of optimizers, Orquestra solves this problem by moving differing parameters into object initialization. Constructed optimizers provide `minimize` method that always accepts only `cost_function` and `initial_params` arguments. Once constructed, optimizers are fully interchangeable.
```
